lib/core-server/src/utils/StoryIndexGenerator.ts

```diff
@@ -56,45 +66,145 @@ export class StoryIndexGenerator {
+      // { title?, of?, imports? }
+      const result = analyze(content);
+
+      const makeAbsolute = (otherImport: Path) =>
+        otherImport.startsWith('.')
+          ? slash(
+              path.join(
+                this.options.workingDir,
+                normalizeStoryPath(path.join(path.dirname(normalizedPath), otherImport))
+              )
             )
-          );
-        })
-      )
-    ).flat();
+          : otherImport;
+
+      const absoluteImports = (result.imports as string[]).map(makeAbsolute);
+      const absoluteOf = result.of && makeAbsolute(result.of);
+
+      let ofTitle: string;
+      const dependencies = [] as StoriesCacheEntry[];
+      this.specifierToCache.forEach((cache) => {
+        const fileNames = Object.keys(cache).filter((fileName) => {
+          return absoluteImports.some((storyImport) => fileName.startsWith(storyImport));
+        });
+        fileNames.forEach((fileName) => {
+          const cacheEntry = cache[fileName];
+          if (cacheEntry && cacheEntry.type === 'stories') {
+            if (fileName.startsWith(absoluteOf) && cacheEntry.entries.length > 0) {
+              ofTitle = cacheEntry.entries[0].title;
+            }
+            dependencies.push(cacheEntry);
+          } else {
+            throw new Error(`Unexpected dependency: ${cacheEntry}`);
+          }
+        });
+      });
+
+      dependencies.forEach((dep) => {
+        dep.dependents.push(absolutePath);
+      });
+
+      const title = result.title || ofTitle || defaultTitle;
+      const name = 'docs';
+      const id = toId(title, name);
+
+      const docsEntry: DocsCacheEntry = {
+        id,
+        title,
+        name,
+        importPath,
+        storiesImports: dependencies.map((dep) => dep.entries[0].importPath),
+        type: 'docs',
+      };
+      return docsEntry;
+    } catch (err) {
+      logger.warn(`🚨 Extraction error on ${relativePath}: ${err}`);
+      throw err;
+    }
   }
 
   async extractStories(specifier: NormalizedStoriesSpecifier, absolutePath: Path) {
     const relativePath = path.relative(this.options.workingDir, absolutePath);
-    const fileStories = {} as StoryIndex['stories'];
-    const entry = this.storyIndexEntries.get(specifier);
+    const entries = [] as StoryIndexEntry[];
     try {
-      const importPath = slash(normalizeStoryPath(relativePath));
+      const normalizedPath = normalizeStoryPath(relativePath);
+      const importPath = slash(normalizedPath);
       const defaultTitle = autoTitleFromSpecifier(importPath, specifier);
       const csf = (await readCsfOrMdx(absolutePath, { defaultTitle })).parse();
-      csf.stories.forEach(({ id, name }) => {
-        fileStories[id] = {
+      const storiesImports = await Promise.all(
+        csf.imports.map(async (otherImport) =>
+          otherImport.startsWith('.')
+            ? slash(normalizeStoryPath(path.join(path.dirname(normalizedPath), otherImport)))
+            : otherImport
+        )
+      );
```

Comment:
Is this for `docsOnly`? I think maybe we should put the whole thing in a `if()` block for clarity.